056-kafka-roller-kraft.md

```diff
@@ -0,0 +1,158 @@
+
+## Motivation
+
+When running in the KRaft mode the controller pods need to be rolled if manually annoted, configuration changes occur or pod is unready. At the moment the 
+existing logic is blocking the Zookeeper to KRaft migration that is being proposed in [PR #93](https://github.com/strimzi/proposals/pull/93) as well as the full implementation of KRaft liveness and readiness checks as described in [proposal #46](https://github.com/strimzi/proposals/blob/main/046-kraft-liveness-readiness.md).
+
+## Proposal
+
+The KafkaRoller behaviour should be unchanged when operating against a ZooKeeper based cluster.
+
+The behaviour when operating against a KRaft cluster is described below.
+
+### Order to restart nodes
+Restart in order:
+1. Unready controller/combined nodes
+2. Ready controller/combined nodes
+3. Active controller (applies to both pure controller and combined node)
+4. Unready broker-only nodes
+5. Ready broker-only nodes
+
+
+### Triggers
+The following are some of the triggers that would restart a KRaft controller or combined node:
+- Pod or its StrimziPodSet is annotated for manual rolling update
+- Pod is unready 
+- Controller's configuration has changed
+
+The triggers for broker remain the same as Zookeeper mode.
+
+### Restart conditions
+The restart conditions that the KafkaRoller considers in different modes are described below.
+
+#### The new quorum check
+The restart conditions include a new check for controllers to verify that restarting the Kafka node does not affect the quorum health.
+The proposed check is:
+- Create admin client connection to the brokers and call `describeMetadataQuorum` API
+- If failed to connect to the brokers, return `UnforceableProblem` which currently results in delay and retry for the pod until the maximum attempts is reached.
+- From the quorum info returned from the admin API, read `lastCaughtUpTimestamp` of each controller. `lastCaughtUpTimestamp` is the last millisecond timestamp at which a replica controller was known to be caught up with the quorum leader.
+- Check the quorum leader id using the quorum info and identify the `lastCaughtUpTimestamp` of the quorum leader
+- Retrieve value of the Kafka property `controller.quorum.fetch.timeout.ms` from the desired configurations specified in the Kafka CR. If this property does not exist in the desired configurations, then use the hard-coded default value for it which is `2000`. The reason for this is explained further in the **NOTE** below.
+- Mark a Kafka node as caught up if `leaderLastCaughtUpTimestamp - replicaLastCaughtUpTimestamp < controllerQuorumFetchTimeoutMs`. This will exclude the controller KafkaRoller is currently considering to restart, because we are trying to check whether the quorum would stay healthy when this controller is restarted.
+- Count each controller node that is caught up, including the leader (`numOfCaughtUpControllers`).
+- Can restart if: `numOfCaughtUpControllers >= ceil((double) totalNumOfControllers / 2)`.
```

Comment:
The calculation is still not completely correct. 
In odds number's case, ex: 3 controllers case, `ceil((double) totalNumOfControllers / 2` -> 2, which is expected.
In even number's case, ex: 4 controllers case, `ceil((double) totalNumOfControllers / 2` -> 2, which is not right. the majority should be 3.

I think it should be `ceil((double) (totalNumOfControllers + 1) / 2`

Reply:
Yes you're right, I'll update it

Reply:
@showuon In this calculation, should "numOfCaughtUpControllers" only include the leader/active controller if we aren't trying to restart the current active leader?

i.e. if we have nodes 0,1,2 and 0 is the active leader, 1 is a caught up follower and 2 is a non-caught up follower, then we can restart 2 but not 0 or 1?

Reply:
Based on discussion in the community call on 5th October I'm going to tweak the wording here